### PR TITLE
Makes the bot more resilient to slow connections/network failures

### PR DESCRIPTION
**Pull request by aloisklink**
Two weekends ago (August 7th-8th), the POAP.art sandbox was lagging quite a bit, with lots of requests timing out.

I made some changes to make the bot more resilient to cases where requests were failing.

Please see each individual commit message for more details on changes made, but as a summary:

- Cancel `paint` calls if they don't finish within 5 seconds (this is what the official POAP.art website does)
- Reconnect the Websocket URL if connection ever fails.
- Removes the `?since=time` HTTP parameter when loading chunks, since the POAP.art website doesn't use it.
  (plus this means that POAP.art's CDN can cache the calls)
- Tidy up Websocket event handling, so that we can ignore `"online"` events instead of `console.log`-ing them.
- Retry loading chunks up to 5 times, in case of error. Otherwise the POAP.art bot was only loading half the canvas.
- Finally, I added an `X-POAP-Art-Bot` HTTP header to all the `paint` calls. The Poap.art haven't asked for this, but I thought it would be polite to add it it in anyway, just in case they want to use it for stats, or if they ever need to block us in an emergency.

I know this is a big pull request, so please let me know if you want me to make separate pull requests for each topic!

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -67,60 +86,87 @@ let enemyArtistsList = new Map()
...
+      img.y <= y && y < img.y + img.h
+    );
+    if (!withinImageRange) {
+      return;
+    }
+
+    const col = drawCtx.getImageData(x, y, 1, 1).data;
+    let expectedColorIdx = approximateColor(col[0], col[1], col[2])
+    let i = (y - img.y) * img.w + x - img.x
+    const wallet = ens ?? address; // use ENS if it exists
```
Comment (aloisklink) [inline index.js:112]: The `??` syntax is called https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator and is a new JavaScript feature.

It was added in Chrome 80 (February 2020), so if people are using MetaMask, their browser should also support this feature.

But we can always replace it with `||` (OR) if think it is worth supporting older browsers (like Internet Explorer) too.

**Comment by RomiRand** [comment]
You've gone through my mess, that's so awesome! It's all reasonable changes and commentary.

Thanks for cleaning up :sweat_smile:

File: index.js (hunk 1 above)
Comment (RomiRand) [reply index.js:112]: I didn't know about this yet, thanks for the info

**Review by RomiRand — COMMENTED** [review]
Didn't think about the `X-POAP-Art-Bot` Header, I like this good practice

**Comment by aloisklink** [comment]
Nah, no worries! It was pretty legible to me. You did all the hard work of reverse engineering the POAP.art API!

I have noticed you program JavaScript like a Python developer, though, but I'm a Python developer too, so it reminds me of when I first learning JavaScript :+1:

If you want to increase legibility, you could always run the code through a program like https://prettier.io/, but it might be overkill for a small project like this!

**Pull request merged**